worker/caasunitprovisioner/worker_test.go

```diff
@@ -0,0 +1,269 @@
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out sending applications change")
+	}
+
+	select {
+	case s.unitChanges <- []string{"gitlab/0"}:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out sending units change")
+	}
+
+	select {
+	case s.containerSpecChanges <- struct{}{}:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out sending container spec change")
+	}
+
+	// TODO(caas) when we have a container broker capable
+	// of provisioning unit containers, check that no unit
+	// is created yet (since there's not yet a spec).
+
+	workertest.CleanKill(c, w)
+	s.applicationGetter.CheckCallNames(c, "WatchApplications")
+	s.unitGetter.CheckCallNames(c, "WatchApplicationUnits")
+	s.unitGetter.CheckCall(c, 0, "WatchApplicationUnits", "gitlab")
+	s.containerSpecGetter.CheckCallNames(c, "WatchContainerSpec", "ContainerSpec")
+	s.containerSpecGetter.CheckCall(c, 0, "WatchContainerSpec", "gitlab/0")
+	s.containerSpecGetter.CheckCall(c, 1, "ContainerSpec", "gitlab/0")
+	s.lifeGetter.CheckCallNames(c, "Life", "Life")
+	s.lifeGetter.CheckCall(c, 0, "Life", "gitlab")
+	s.lifeGetter.CheckCall(c, 1, "Life", "gitlab/0")
+}
+
+func (s *WorkerSuite) TestWatchApplicationDead(c *gc.C) {
+	w, err := caasunitprovisioner.NewWorker(s.config)
+	c.Assert(err, jc.ErrorIsNil)
+	defer workertest.CleanKill(c, w)
+
+	s.lifeGetter.life = life.Dead
+	select {
+	case s.applicationChanges <- []string{"gitlab"}:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out sending applications change")
+	}
+
+	select {
+	case s.unitChanges <- []string{"gitlab/0"}:
+		c.Fatal("unexpected watch for units")
+	case <-time.After(coretesting.ShortWait):
+	}
+
+	workertest.CleanKill(c, w)
+	s.unitGetter.CheckNoCalls(c)
+}
+
+func (s *WorkerSuite) TestWatchUnitDead(c *gc.C) {
+	w, err := caasunitprovisioner.NewWorker(s.config)
+	c.Assert(err, jc.ErrorIsNil)
+	defer workertest.CleanKill(c, w)
+
+	select {
+	case s.applicationChanges <- []string{"gitlab"}:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out sending applications change")
+	}
+
+	s.lifeGetter.life = life.Dead
+	select {
+	case s.unitChanges <- []string{"gitlab/0"}:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out sending units change")
+	}
+
+	workertest.CleanKill(c, w)
+	s.containerSpecGetter.CheckNoCalls(c)
+}
+
+func (s *WorkerSuite) TestRemoveApplicationStopsWatchingApplication(c *gc.C) {
+	w, err := caasunitprovisioner.NewWorker(s.config)
+	c.Assert(err, jc.ErrorIsNil)
+	defer workertest.CleanKill(c, w)
+
+	select {
+	case s.applicationChanges <- []string{"gitlab"}:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out sending applications change")
+	}
+
+	s.lifeGetter.SetErrors(errors.NotFoundf("application"))
+	select {
+	case s.applicationChanges <- []string{"gitlab"}:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out sending applications change")
+	}
+
+	workertest.CheckKilled(c, s.unitGetter.watcher)
```

Comment:
check the spec watcher also since it should be killed along with the unit watcher when the app goes away?

Reply:
no, because we never send a units change - hence there are no unit workers running